### ISSUE TITLE
Add state query functions / state change messages

### ISSUE THREAD

**Issue by Robert-M-Muench** [Evertras/bubble-table#60]
Things I want to check outside the table:
* Which column is the table sorted by?
* Is a filter active?
* What is the current filter string?

Possible tea.Msg:
* filtering started
* filtering changed
* filtering ended
* row changed
* row selected
* etc.

**Comment by Evertras** [comment]
I think adding events is reasonable.  I'll need to toy with this a bit.  I think adding the other checks is also reasonable but potentially a separate issue to plug some gaps here.

**Comment by Evertras** [comment]
Some basic queries have been added as listed above, but I'm currently struggling to figure out a nice way to add events that fits well into the Bubble Tea architecture without opening a massive can of worms.  Still toying with it when I have time...

**Comment by Robert-M-Muench** [comment]
What problem do you see with defining some own type.Msg messages?

**Comment by Evertras** [comment]
I think having some event types that could be switched on makes sense, but the issue is in actually returning the value.  To fit into the Bubble Tea architecture as smoothly as possible, it returns `Model, tea.Cmd`.  We would have to wrap the events as `tea.Cmd`, [which feels at odds with the intent of Cmd](https://github.com/charmbracelet/bubbletea/blob/master/tea.go#L50-L57).  It wants to send stuff through `Update`, but that's only returning `Cmd`, so I'm trying to figure out the best way to go here.

I'm also slightly concerned with potential infinite loops.  You receive an event that something about the table has changed, so you change something about the table, which then sends something about the table changing... for example, updating row data when a row is updated.  I think the obvious loops will be easy to avoid but it feels like a slippery slope, and it's not a pattern I see in the main architecture.

**Comment by Robert-M-Muench** [comment]
The official way to send a `Msg` is by `Program.Send`, which then enters the `Update` function at some point. See:
https://pkg.go.dev/github.com/charmbracelet/bubbletea#Program.Send This is the access point to the internal message queue.

IIRC `Msg`s are collected and fed into the `Update` function in random order and can happen in parallel.

One idea would be to collect all events during the bubble-tables `Update` function and send them as `Msg` before exiting the function, communicating them from a stable state.

Sending events as soon as they are triggered might be a bit wilder but more in line with the event concept of "something happened." 

However, if a user makes changes during change events, this requires explicit synchronization.

**Comment by Robert-M-Muench** [comment]
Here is an example using channels: https://github.com/charmbracelet/bubbletea/discussions/305 that might give some inspiration.

**Comment by Evertras** [comment]
Sorry for being a bit slow on this, having some discussions and some considerations.

You are correct that the `Msg`s are not guaranteed to be sent in order, which is one major problem.  It's actually quite simple to batch all the events that have happened in a single `Update` and return them as a single batched `tea.Cmd` to feed the messages in sequentially, but the problem is that these events are not generally going to be together anyway.

As a thought experiment, consider the filter input.  The user hits Enter, which starts the filter.  A `FilterActivated` event is generated, returned as a `tea.Cmd`, and added to the general pool of `StuffHappened` in Bubble Tea, waiting to be asynchronously returned.  At the exact same moment, an API refresh comes back with fresh data.  Bubble Tea decides to send this in first, because the order isn't guaranteed and either has a chance of going through.  The user has decided that receiving data should `Blur()` the filter, so they programmatically `Blur` it.  This creates a `FilterDeactivated` event which returns immediately, adding it to the async `StuffHappened` pool.  Bubble Tea then goes to 'pick' the next message to send, and randomly chooses (or rather, Go chooses via channel, but semantics) the `FilterDeactivated` message.  Then afterwards the `FilterActivated` message finally gets sent, and if the code checks it would actually find that the filter is not actually currently active despite the event...

This has the potential to do really wacky things to state in general.  I don't think doing this in an async way is safe.  There's also the issue that just got pointed out to me of the fact that you'd need to figure out which table this is associated with in the first place, which is going to be a pain in tracking things and making connections again within the `Update`s.

As an alternative, I think something like this may be more doable:

```golang
m.myTable, cmd := m.myTable.Update()

for _, event := range m.myTable.UpdateEvents() {
  switch event.(type) {
    // ... react to events synchronously here ...
  }
}
```

Basically `UpdateEvents` would just be a list of stuff that changed during the last `Update` call.  Keeping it synchronous like this gets rid of all the nasty race conditions and infinite loop traps that concern me.  It's not quite as elegant, but I think it's a lot safer/simpler to reason about.

**Comment by Robert-M-Muench** [comment]
Yes, that was my idea too while reading the first part of your answer.

Explicitly _"polling"_ the events make things much more simpler and useable in the large. Some notes / ideas:
* CheckForEvent() returns true/false and keeps the event in the queue
* Events should be ordered in time, so when using range I should get the oldest event first
* Maybe a way to turn the recording of some events to avoid flooding the event queue, like _I'm not interested in any cursor movement events_

**Comment by Evertras** [comment]
I'm a little wary of trying to recreate a full event-based framework here.  With this kind of functionality it's not just a table, it's a full event system, and I think that's out of scope.  If you wanted to include other events from other sources, for example, you would have to recreate this for other event sources or somehow jam them through the table's queue.

I think the furthest this can reasonably go is having something like the code snippet above, and it's just a single well-ordered slice of events that have happened during that single `Update` call.  This slice is dropped/recreated every time `Update` is called, so it's the caller's responsibility to deal with events immediately after the `Update` call.  For users that are not interested in events, there's no change or queue leak to worry about.  For users that are interested in simple events, they can do a quick raw `range` through as in the code above.  For users that are doing more complicated things with events, they can feed the slice into some other system.

Does this sound reasonable?

**Comment by Robert-M-Muench** [comment]
My ideas were mean to be only within the table context. So, nothing leaking to the outside of it.

However, let's start with your idea and see how things work. Sounds reasonable.

**Comment by Robert-M-Muench** [comment]
Any progress on this one?

I'm currently facing the case, where I need to load some data into a sub-table when a new line is highlighted in the main table. Here the *row changed* event would be helpful.

**Comment by Evertras** [comment]
Let me take a shot at this in the next few days. Admittedly been dragging my feet a bit as I’m concerned about scope but I think the use case above will be a good limiter there.

**Comment by Robert-M-Muench** [comment]
Sending state information messages is a proven pattern, and I think a table is one of the most complex widgets in applications.

Hence, somehow this state information is needed. Either by providing an extensive API where I can query all kinds of things or collect them as you suggested.

(BTW: Bubble-Table is fantastic, and it's great to see activity and progress.)

**Comment by Evertras** [comment]
Glad you find it helpful!  Agreed on the complexity, even with the visualizations alone.

To answer an earlier question regarding on the best way to do this right now:

> I'm currently facing the case, where I need to load some data into a sub-table when a new line is highlighted in the main table. Here the row changed event would be helpful.
> What's the best practice to do this currently?

The best way to do this right now before I try adding some event collection would be to track the highlighted row, then after calling `table.Update` you would check to see if the highlighted row is different.  This could trigger your sub-table load for now.  I think chasing the `KeyMsg` will be painful.

**Comment by Robert-M-Muench** [comment]
`TotalRows()` returns an `int`, `VisibleIndices()` two ints, `WithHighlightedRow(index int)` takes an int based index too, but there is no way to get the current highlighted row's index.

`HighlightedRow()` returns row data, which is not nice to compare because I need to write a custom compare function.

**Comment by Evertras** [comment]
That feels like an oversight on my part.  I'll take a look at that as well when I get some time.  Currently a bit eaten by day job, so it may need to wait until the weekend.

**Comment by Evertras** [comment]
I've added that along with horizontal scrolling offset and whether the table is focused or not, and released this as `v0.13.4`.  Also as a note, your original list of what you'd like to query is implemented, so let me know if there are any other internals you'd like to get at.

Going to finally take a crack at collecting events...

**Comment by Evertras** [comment]
@Robert-M-Muench could you take a look at #98?  Specifically here: https://github.com/Evertras/bubble-table/pull/98/files#diff-4e8403e8772b3e15cb558060b5c8cb8fd2a07b263a1836da16548c8318fcf959R150-R165

Does that usage seem consistent with what you were thinking and/or would be useful?  I want to double check before merging/releasing this as I think this will lock us into a very particular pattern the moment I merge to main.

**Comment by Robert-M-Muench** [comment]
This change already looks great! It's a generic pattern that will be expandable without breaking anything.

> These will NOT be generated when code programmatically changes values in the table.

Not sure about this. What do you mean by * programmatically*: Directly manipulating the table's state or using the table's API to control the state?

However, NOT generating events when a change happens, while events get generated when the user does it, results in an inconsistency. If the table changes, an event should be generated. It doesn't matter if it was the user or programmatically (assuming we do not mean any direct internal state hacking).

If we want to distinguish the two cases, I think it's better to add a *programmatic* flag to each event and set it. With this, my code doesn't have to handle different events/cases, and I can easily distinguish between the two sources.

**Comment by Evertras** [comment]
> Not sure about this. What do you mean by * programmatically*: Directly manipulating the table's state or using the table's API to control the state?

As in using the API, to clarify.  There should be no way to manipulate the table's state directly, or at least this is undefined/unsupported behavior if someone were to try and get clever this way through editing some slices/refs that I missed.

I would argue that in this case, the scope of the events is important to consider when talking about consistency.  Here the events are scoped to "a thing that happened due to user input when you called `myTable.Update()`".  This is currently a complete black box to the rest of the code; some message went in, and maybe some internal state changed on the table that you would like to know about.  With this change, we start to expose those state changes through events.  If you were already changing the table yourself directly through the API, you already have the hook for doing something you want to be doing, such as triggering some API call for external data, etc.

To me, it's about the source of the event.  In `Update`, the table is generating a change event, and the rest of the code flow would like to react to it.  Outside of `Update`, something else in the code flow is generating the change event, and the table is asked to react to that change via the API.

This is also for practical reasons, which I think manifest from the question of source above.  We need to clear the `updated` event list for every `Update()` call, which is relatively straightforward.  When would we clear the events for an API change that may have happened at any point in the program flow?  Do we ask the user to do so?  That feels quite leaky, and potentially dangerous.  Do we rely on the `Update` clear anyway?  This feels messy, because then ordering each `Update` for all components becomes important in also-leaky ways.  Should the API return events directly?  I can kind of see that, but that's an incredibly breaking change at this point and it feels redundant to me for most operations.

Additionally, no other components will be following this system; the table will be the only one with these kinds of events.  Do we ask other components to follow suit?  I think trying to figure out some kind of event-based framework to overlay on top of various components (including the table, using `Update` events to help feed higher level events into the framework) is more what you're driving for here.

Also to be crystal clear, I appreciate the discussion, even if I think we disagree on what's in scope! ❤️

**Comment by Robert-M-Muench** [comment]
I see your point and think there is no right or wrong to both of our views.

I try to keep my Update handling code (get state of the table and do something in the app) in one place. Hence it would be good to have access to all events since the last `table.Update()` call.

That `table.Update()` clears the collected events makes sense because that's a simple rule.

Manipulating the table state through an API (inbound) from different places, while the table collects all these events, allows me to get a summary (outbound) of what happened since the last time table handled the user input in one place. 

This approach gives a precise central location where you know the information you get exactly. 

Flagging the API programmatically generated events would allow me to skip them.

I think that would combine both of our use-cases/views and gives a straightforward but highly flexible solution.

The pattern to collect status events in the component and making these available between Update() calls is IMO a handy thing, and many more components should support it.

**Comment by Evertras** [comment]
I started playing with this idea, but ran into a tricky inconsistency if we assume that events are cleared on the `Update` loop.  Goofy code below:

```
func Update() {
  if something {
    myTable = myTable.PageDown()
  }

  myTable = myTable.Update()

  if somethingElse {
    myTable = myTable.PageUp()
  }
}
```

So here's the question: when do you check for events?

```
func Update() {
  // A

  if something {
    myTable = myTable.PageDown()
  }

  // B

  myTable = myTable.Update()

  // C

  if somethingElse {
    myTable = myTable.PageUp()
  }

  // D
}
```

If you check at D, then you miss out on the `PageDown` event above because the table has cleared it in `Update`.  If you check at C or A, you also miss out on the `PageUp` event for the same reason.  This is completely not obvious, but the trickiness doesn't end there.  If you try to go for B because you know the underlying system (which is already an overly leaky design), you get into a potentially very strange state where you can't actually react to the events for some indeterminate delay because you can't be sure when the next `Update` is called.  Unless you then also return some dummy command from the table's Update which might then trigger the outer Update again, which only makes the leakiness worse and potentially creates an infinite loop in the worst case.

I don't mind the user-generated events as in the PR, but I'm continuing to see very nasty complications when we try to go outside of that.  For now I think I'm going to limit the scope to events within Update only, as that's still solving a useful problem while being relatively easy to use.

**Comment by Robert-M-Muench** [comment]
My initial idea to use *official* Bubbletea cmds/messages was based on using as much existing design as possible without introducing a parallel infrastructure. With this, the table would return all collected events, Bubbletea would handle them as a result of the *Update()* call chain, and the app handles it or not like other messages.

Your idea is simpler and more pragmatic while being a bit *out of bound* with the rest. Hence, I think we shouldn't make things too complicated because of trying to find a perfect solution.

Some ideas:

1. Push the event-log clearing action to the user. I can see quite some advantages in doing this. Combined with pushing all events to the event-log (API & user), I could record a full replayable event stream.
2. Clear on entering *Update()* and have the user get the event-log (as copy) and manage the event-log lifetime themselves. In your case, I would get the event-log several times and stitch the parts together.
3. Add an *Update* event (maybe even EnterUpdate/ExitUpdate), so I can spot where in the event-log the update happened. I think that makes sense anyway.

Overall, I don't think it's necessary to sync event-log clearing and *Update()* calls. The table generates a flow of event-log entries, then I query them and analyze and act however it makes sense.

**Comment by Evertras** [comment]
The more we dig into this, the more I'm getting convinced that this is not in the scope of the table component itself, but rather as a pattern that should be implemented/supported at the Bubbletea level itself to actually be valuable.  The cmds/messages are not intended as events (which I confirmed with a Bubbletea dev, and we hit some async madness real quick when we tried to work through the idea), which puts us in the out of bounds area anyway.  So writing up a complicated, custom implementation for this table seems like a lot of overkill and extra leaky complexity for something that can be done by doing some basic wrapping from outside the table.

At this point I think I'm going to have to put my foot down and say that programmatic events are not going to be implemented.  I'm willing to do the update events, but I don't want to go further than that.

**Comment by Robert-M-Muench** [comment]
> The cmds/messages are not intended as events (which I confirmed with a Bubbletea dev, and we hit some async madness real quick when we tried to work through the idea)

That's interesting. Because I think that's what cmds/messages exactly look like. Did you have your discussion public so that I could join?

I agree that a component shouldn't implement fundamental parts like an event system but use the existing facilities from the framework.

A fair point regarding the programmatic events: getting the user events is a big step ahead and let's see how it works out. I have an immediate use-case for them :-)

**Comment by Evertras** [comment]
Released `v0.14.0` with the highlight change event.  Apologies for slowness lately, day job is eating my dev energy lately.

The Charm slack is here if you'd like to join: https://charm.sh/slack/ and the conversation we had was here: https://charmbracelet.slack.com/archives/CQ2E9CGTV/p1650375996352089 I think it would be interesting to revive, maybe in a new thread so others can chime in as well.
